addons/zylann.hterrain/native/src/quad_tree_lod.cpp

```diff
@@ -0,0 +1,245 @@
+#include "quad_tree_lod.h"
+
+namespace godot {
+
+void QuadTreeLod::_register_methods() {
+	register_method("set_callbacks", &QuadTreeLod::set_callbacks);
+	register_method("get_lod_count", &QuadTreeLod::get_lod_count);
+	register_method("get_lod_factor", &QuadTreeLod::get_lod_factor);
+	register_method("compute_lod_count", &QuadTreeLod::compute_lod_count);
+	register_method("set_split_scale", &QuadTreeLod::set_split_scale);
+	register_method("get_split_scale", &QuadTreeLod::get_split_scale);
+	register_method("clear", &QuadTreeLod::clear);
+	register_method("create_from_sizes", &QuadTreeLod::create_from_sizes);
+	register_method("update", &QuadTreeLod::update);
+	register_method("debug_draw_tree", &QuadTreeLod::debug_draw_tree);
+}
+
+void QuadTreeLod::set_callbacks(Ref<FuncRef> make_cb, Ref<FuncRef> recycle_cb, Ref<FuncRef> vbounds_cb) {
+	_make_func = make_cb;
+	_recycle_func = recycle_cb;
+	_vertical_bounds_func = vbounds_cb;
+}
+
+int QuadTreeLod::get_lod_count() {
+	// TODO make this a count, not max
+	return _max_depth + 1;
+}
+
+int QuadTreeLod::get_lod_factor(int lod) {
+	return 1 << lod;
+}
+
+int QuadTreeLod::compute_lod_count(int base_size, int full_size) {
+	int po = 0;
+	while (full_size > base_size) {
+		full_size = full_size >> 1;
+		po += 1;
+	}
+	return po;
+}
+
+// The higher, the longer LODs will spread and higher the quality.
+// The lower, the shorter LODs will spread and lower the quality.
+void QuadTreeLod::set_split_scale(real_t p_split_scale) {
+	real_t MIN = 2.0f;
+	real_t MAX = 5.0f;
+
+	// Split scale must be greater than a threshold,
+	// otherwise lods will decimate too fast and it will look messy
+	if (p_split_scale < MIN)
+		p_split_scale = MIN;
+	if (p_split_scale > MAX)
+		p_split_scale = MAX;
+
+	_split_scale = p_split_scale;
+}
+
+real_t QuadTreeLod::get_split_scale() {
+	return _split_scale;
+}
+
+// Intention is to only clear references to children
+void QuadTreeLod::_clear_children(unsigned int index) {
+    Quad* quad = _get_node(index);
+    if (quad->has_children()) {
+        _recycle_children(quad->first_child);
+        quad->first_child = NO_CHILDREN;
+    }
+}
+
+// Returns the index of the first_child. Allocates from _free_indices.
+unsigned int QuadTreeLod::_allocate_children() {
+    if (_free_indices.size() == 0) {
+        unsigned int i0 = static_cast<unsigned int>(_tree.size());
+        _tree.resize(i0 + 4);
+        return 4 * i0;
+    } else {
+        unsigned int i0 = _free_indices[_free_indices.size() - 1];
+        _free_indices.pop_back();
+        return 4 * i0;
+    }
+}
+
+// Pass the first_child index, not the parent index. Stores back in _free_indices.
+void QuadTreeLod::_recycle_children(unsigned int i0) {
+    // Debug check, there is no use case in recycling a node which is not a first child
+    CRASH_COND(i0 % 4 != 0);
+
+    for (int i = 0; i < 4; ++i) {
+        _tree[i0 + i].init();
+    }
+
+    _free_indices.push_back(i0 / 4);
```

Comment:
And why divide by 4? Seems like you are working with a different type of index here, but it's really unnecessary. You can totally store actual indexes to first children in there, still without storing the 3 others

Reply:
We've adjusted this. _free_indices is still only 1/4th the size, and it stores only the multiples of 4: [ 0, 4, 8, 16].